[tflite] add int8 input/output to label_image

freedomtan opened the pull request:
More and more models, such as [MobilenetV3's EdgeTPU ones](https://github.com/tensorflow/models/tree/master/research/slim/nets/mobilenet), are using [post-training full integer quantization](https://www.tensorflow.org/lite/performance/post_training_quantization#full_integer_quantization_of_weights_and_activations). With this patch, I can get reasonable results.

```
./label_image_int8 -m mobilenet_edgetpu_224_1.0_int8.tflite
Loaded model mobilenet_edgetpu_224_1.0_int8.tflite
resolved reporter
INFO: Initialized TensorFlow Lite runtime.
invoked
average time: 15.363 ms
0.867188: 653 military uniform
0.0390625: 835 suit
0.015625: 458 bow tie
0.0078125: 907 Windsor tie
0.00390625: 716 pickelhaube
```

freedomtan committed:
[tflite] add int8 input/output to label_image

More and more models, such as MobilenetV3's EdgeTPU ones, are using
post-training full integer quantization. With this patch, I can get
reasonable results.

./label_image_int8 -m mobilenet_edgetpu_224_1.0_int8.tflite
Loaded model mobilenet_edgetpu_224_1.0_int8.tflite
resolved reporter
INFO: Initialized TensorFlow Lite runtime.
invoked
average time: 15.363 ms
0.867188: 653 military uniform
0.0390625: 835 suit
0.015625: 458 bow tie
0.0078125: 907 Windsor tie
0.00390625: 716 pickelhaube

freedomtan committed:
clean up

jianlijianli commented on tensorflow/lite/examples/label_image/get_top_n_impl.h:
```diff
@@ -30,19 +30,29 @@ extern bool input_floating;
+        value = prediction[i];
+        break;
+      case kTfLiteInt8:
+        value = (prediction[i] + 128) / 256.0;
```
Why is this 256 while the uint8 case is 255?

freedomtan replied:
This is to dequantize values in the final op. For uint8 cases, as far as I can tell, most models generated with quantized-aware training have value between 0 and 255.  

The last layer of the int8 [Mobilenet V3 EdgeTPU model](https://storage.cloud.google.com/mobilenet_edgetpu/checkpoints/mobilenet_edgetpu_224_1.0.tgz) I used to test this, has scale = 0.00390625 = 1 / 256.

freedomtan commented on tensorflow/lite/examples/label_image/bitmap_helpers_impl.h:
```diff
@@ -82,10 +82,19 @@ void resize(T* out, uint8_t* in, int image_height, int image_width,
+        out[i] = output[i] - 128;
+        break;
+      case kTfLiteUInt8:
+        out[i] = output[i];
```
Sounds right.

freedomtan committed:
address review comments

1. add explicit cast back
2. change int to TfLiteType

gbaned commented:
@srjoglekar246 Can you please take a look on this PR? Thanks!

rthadur commented:
Here are internal failures , can you please check 
`In file included from ./third_party/tensorflow/lite/examples/label_image/get_top_n.h:19:
./third_party/tensorflow/lite/examples/label_image/get_top_n_impl.h:33:16: error: unknown type name 'TfLiteType'
               TfLiteType input_type) {
               ^
./third_party/tensorflow/lite/examples/label_image/get_top_n.h:27:16: error: unknown type name 'TfLiteType'
               TfLiteType input_type);
               ^
./third_party/tensorflow/lite/examples/label_image/get_top_n.h:31:69: error: unknown type name 'TfLiteType'
                               std::vector>*, TfLiteType);
                                                                    ^
./third_party/tensorflow/lite/examples/label_image/get_top_n.h:30:15: error: explicit instantiation of 'get_top_n' does not refer to a function template, variable template, member function, member class, or static data member
template void get_top_n(float*, int, size_t, float,
              ^
./third_party/tensorflow/lite/examples/label_image/get_top_n.h:34:33: error: unknown type name 'TfLiteType'
                                TfLiteType);
                                ^
./third_party/tensorflow/lite/examples/label_image/get_top_n.h:32:15: error: explicit instantiation of 'get_top_n' does not refer to a function template, variable template, member function, member class, or static data member
template void get_top_n(int8_t*, int, size_t, float,
              ^
./third_party/tensorflow/lite/examples/label_image/get_top_n.h:37:34: error: unknown type name 'TfLiteType'
                                 TfLiteType);
                                 ^
./third_party/tensorflow/lite/examples/label_image/get_top_n.h:35:15: error: explicit instantiation of 'get_top_n' does not refer to a function template, variable template, member function, member class, or static data member
template void get_top_n(uint8_t*, int, size_t, float,
              ^
In file included from ./third_party/tensorflow/lite/examples/label_image/get_top_n.h:19:
./third_party/tensorflow/lite/examples/label_image/get_top_n_impl.h:44:12: error: use of undeclared identifier 'kTfLiteFloat32'
      case kTfLiteFloat32:
           ^
./third_party/tensorflow/lite/examples/label_image/get_top_n_impl.h:47:12: error: use of undeclared identifier 'kTfLiteInt8'
      case kTfLiteInt8:
           ^
./third_party/tensorflow/lite/examples/label_image/get_top_n_impl.h:50:12: error: use of undeclared identifier 'kTfLiteUInt8'
      case kTfLiteUInt8:
           ^
11 errors generated.

./third_party/tensorflow/lite/examples/label_image/get_top_n_impl.h:33:16: error: unknown type name 'TfLiteType'
               TfLiteType input_type) {
               ^
./third_party/tensorflow/lite/examples/label_image/get_top_n_impl.h:44:12: error: use of undeclared identifier 'kTfLiteFloat32'
      case kTfLiteFloat32:
           ^
./third_party/tensorflow/lite/examples/label_image/get_top_n_impl.h:47:12: error: use of undeclared identifier 'kTfLiteInt8'
      case kTfLiteInt8:
           ^
./third_party/tensorflow/lite/examples/label_image/get_top_n_impl.h:50:12: error: use of undeclared identifier 'kTfLiteUInt8'
      case kTfLiteUInt8:`

srjoglekar246 commented:
I think we need to add a BUILD dependency on `tensorflow/lite/c:common` & import `common.h` for this to BUILD properly using bazel. @freedomtan can you take a look?

freedomtan commented:
@srjoglekar246 I cannot reproduce what @rthadur said on my Mac (with either Python 2 or 3). I'll add the dependency you suggest later anyway.

freedomtan committed:
avoid problem reported by Google's internal test

freedomtan committed:
Merge branch 'master' into label_image_int8

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -36,6 +36,7 @@ limitations under the License.
 #include <vector>
 
 #include "absl/memory/memory.h"
+#include "tensorflow/lite/c/common.h"
```
We don't need this here, since `model.h` imported in `label_image.h` already indirectly includes it.

We need to add it to `get_top_n.h` & `get_top_n_impl.h` - see the error message posted by @gbaned .

freedomtan replied:
Thanks. I'll add it to `get_top_n_impl.h` (since `get_top_n.h` includes  `get_top_n_impl.h` already).

srjoglekar246 reviewed (changes requested):
You probably won't be able to reproduce the error, mainly because they came from our on-device tests. Those usually require all dependencies to be explicit at BUILD time :-)

freedomtan committed:
including "tensorflow/lite/c/common.h" in get_top_n_impl.h

move including "tensorflow/lite/c/common.h" from
label_image.cc to get_top_n_impl.h

srjoglekar246 commented on tensorflow/lite/examples/label_image/BUILD:
```diff
@@ -28,6 +28,7 @@ cc_binary(
     }),
     deps = [
         ":bitmap_helpers",
+        "//tensorflow/lite/c:common",
```
We need this in label_image_test too.

freedomtan replied:
oops, thanks.

freedomtan committed:
add `//tensorflow/lite/c:common` to label_image_test

srjoglekar246 commented on tensorflow/lite/examples/label_image/get_top_n_impl.h:
```diff
@@ -30,19 +32,29 @@ extern bool input_floating;
 void get_top_n(T* prediction, int prediction_size, size_t num_results,
                float threshold, std::vector<std::pair<float, int>>* top_results,
-               bool input_floating) {
+               TfLiteType input_type) {
```
This is causing issues in `label_image_test.cc`, where some tests pass in a bool at parameter no. 6. So you will probably have to modify that file.

freedomtan replied:
Thanks. Fixed.

freedomtan committed:
fix label_image_test

make
```
bazel test //tensorflow/lite/examples/label_image:label_image_test
```
work again.

freedomtan commented:
@yifeif May I know the reason of rollback so that I can fix it?

yifeif commented:
Thanks for checking @freedomtan! It broke one of the internal address sanitizer test. @srjoglekar246, do you mind working with @freedomtan to get this change in again? Thank you :)

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image_test.cc:
```diff
@@ -37,15 +37,15 @@ TEST(LabelImageTest, GraceHopper) {
-  ASSERT_EQ(output[0], 0x15);
-  ASSERT_EQ(output[214 * 214 * 3 - 1], 0x11);
+  ASSERT_EQ(output[0], 0x0);
+  ASSERT_EQ(output[214 * 214 * 3 - 1], 0x0);
```
The issue is with these lines. The input actually *is* uint8 (and not float). Line 31 initializes the `Settings` object to its default values, which assumes `TfLiteType input_type = kTfLiteFloat32;`. So these values `0x15` & `0x11` should not be changed, we should rather set `s.input_type = kTfLiteUInt8` (after importing `tensorflow/lite/c/common.h`).

freedomtan replied:
Thanks. Will fix it.

freedomtan replied:
Since it was merged. Open a new issue (#37502) seems cleaner to me.
